### PR TITLE
Update Docs/Model setup pages + fix bug with `callsite=TendencyCallsite()` in Callbacks

### LINKED ISSUE #3275: Doc page for Callbacks outdated and does not appear under `Docs/Model Setup`
The `callbacks.md` page 

https://github.com/CliMA/Oceananigans.jl/blob/2328d54ba4956f94af3c93f02b7aad4baf454049/docs/src/model_setup/callbacks.md

does not appear in the docs. That is, it's not in this list:

https://github.com/CliMA/Oceananigans.jl/blob/2328d54ba4956f94af3c93f02b7aad4baf454049/docs/make.jl#L94-L112

Furthermore, the `callbacks.md` page needs some TLC (=tender love and care) since it seems bit outdated.

cc @glwagner, @jagoosw, @johnryantaylor

### LINKED ISSUE #3456: Callback with `callsite=TendencyCallsite()` doesn't work
E.g.,

```Julia
using Oceananigans
model = NonhydrostaticModel(grid=RectilinearGrid(size=(1, 1, 1), extent = (1, 1, 1)))
simulation = Simulation(model, Δt = 1, stop_iteration = 3)
test_sim(sim) = @info "Sim? $(isa(sim, Oceananigans.Simulation))"
test_model(model) = @info "Model? $(isa(model, Oceananigans.AbstractModel))"
simulation.callbacks[:inter] = Callback(test_sim)
simulation.callbacks[:intra] = Callback(test_model; callsite=TendencyCallsite())
run!(simulation)
```

gives

```Julia
[ Info: Initializing simulation...
[ Info: Sim? true
[ Info:     ... simulation initialization complete (89.746 ms)
[ Info: Executing initial time step...
[ Info:     ... initial time step complete (5.837 seconds).
[ Info: Sim? true
[ Info: Sim? true
[ Info: Simulation is stopping after running for 5.948 seconds.
[ Info: Model iteration 3 equals or exceeds stop iteration 3.
[ Info: Sim? true
```

### PR DESCRIPTION
**Pull request by navidcy**
This PR

- updates Docs/Grids page
- updates Docs/Architectures page
- updates Docs/Number types page
- resurrects the Docs/Callbacks page

Closes #3275
Closes #3456

### CODE REVIEW

**Comment by navidcy** [comment]
Are the `state_callbacks` something completely deprecated?

```Julia
julia> using Oceananigans

julia> model = NonhydrostaticModel(grid=RectilinearGrid(size=(1, 1, 1), extent=(1, 1, 1)))
NonhydrostaticModel{CPU, RectilinearGrid}(time = 0 seconds, iteration = 0)
├── grid: 1×1×1 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×3×3 halo
├── timestepper: QuasiAdamsBashforth2TimeStepper
├── tracers: ()
├── closure: Nothing
├── buoyancy: Nothing
└── coriolis: Nothing

julia> function modify_tracer(model, params)
           model.timestepper.Gⁿ[params.c] .+= params.δ
       end
modify_tracer (generic function with 1 method)

julia> model.state_callbacks[:modify_u] = Callback(modify_tracer, IterationInterval(1), (c = :u, δ = 1))
ERROR: MethodError: no method matching Callback(::typeof(modify_tracer), ::IterationInterval, ::NamedTuple{(:c, :δ), Tuple{Symbol, Int64}})

Closest candidates are:
  Callback(::F, ::S, ::P, ::CS) where {P, F, S, CS}
   @ Oceananigans ~/Research/OC10.jl/src/Simulations/callback.jl:8
  Callback(::Any, ::Any; parameters, callsite)
   @ Oceananigans ~/Research/OC10.jl/src/Simulations/callback.jl:46
  Callback(::Oceananigans.OutputWriters.WindowedTimeAverage, ::Any; kw...)
   @ Oceananigans ~/Research/OC10.jl/src/Simulations/callback.jl:68
  ...

Stacktrace:
 [1] top-level scope
   @ REPL[56]:1
```

**Comment by navidcy** [comment]
OK, I see the `callsite` kwarg which can be `TendencyCallsite`. But it doesn't seem to work..

```julia
using Oceananigans
model = NonhydrostaticModel(grid=RectilinearGrid(size=(1, 1, 1), extent = (1, 1, 1)))
simulation = Simulation(model, Δt = 1, stop_iteration = 3)
test_sim(sim) = @info "Sim? $(isa(sim, Oceananigans.Simulation))"
test_model(model) = @info "Model? $(isa(model, Oceananigans.AbstractModel))"
simulation.callbacks[:inter] = Callback(test_sim)
simulation.callbacks[:intra] = Callback(test_model; callsite=TendencyCallsite())
run!(simulation)
```

```Julia
[ Info: Initializing simulation...
[ Info: Sim? true
[ Info:     ... simulation initialization complete (13.300 ms)
[ Info: Executing initial time step...
[ Info:     ... initial time step complete (1.012 ms).
[ Info: Sim? true
[ Info: Sim? true
[ Info: Simulation is stopping after running for 18.896 ms.
[ Info: Model iteration 3 equals or exceeds stop iteration 3.
[ Info: Sim? true
```

**Comment by jagoosw** [comment]
Is the `TendancyCallsite` and update state ones still not working?

**Comment by navidcy** [comment]
Doesn’t seem to

**Comment by jagoosw** [comment]
Found the problem I think:

https://github.com/CliMA/Oceananigans.jl/blob/c3151d616de2cb8d3cedae0bbf452004d4cc1065/src/Simulations/run.jl#L109C74-L109C87

I think this should be:
`... if cb.callsite isa ModelCallsite)`

**Comment by navidcy** [comment]
```Julia
using Oceananigans
model = NonhydrostaticModel(grid=RectilinearGrid(size=(1, 1, 1), extent = (1, 1, 1)))
simulation = Simulation(model, Δt = 1, stop_iteration = 3)
test_sim(sim) = @info "Sim? $(isa(sim, Oceananigans.Simulation))"
test_model(model) = @info "Model? $(isa(model, Oceananigans.AbstractModel))"
simulation.callbacks[:inter] = Callback(test_sim)
simulation.callbacks[:intra] = Callback(test_model; callsite=TendencyCallsite())
run!(simulation)
```

### before this PR

```Julia
[ Info: Initializing simulation...
[ Info: Sim? true
[ Info:     ... simulation initialization complete (89.746 ms)
[ Info: Executing initial time step...
[ Info:     ... initial time step complete (5.837 seconds).
[ Info: Sim? true
[ Info: Sim? true
[ Info: Simulation is stopping after running for 5.948 seconds.
[ Info: Model iteration 3 equals or exceeds stop iteration 3.
[ Info: Sim? true
```

### after this PR


```Julia
[ Info: Initializing simulation...
[ Info: Sim? true
[ Info:     ... simulation initialization complete (8.364 ms)
[ Info: Executing initial time step...
[ Info: Model? true
[ Info: Model? true
[ Info:     ... initial time step complete (962.871 ms).
[ Info: Sim? true
[ Info: Model? true
[ Info: Sim? true
[ Info: Model? true
[ Info: Simulation is stopping after running for 990.074 ms.
[ Info: Model iteration 3 equals or exceeds stop iteration 3.
[ Info: Sim? true
```

**Comment by glwagner** [comment]
nice work

File: docs/src/model_setup/architecture.md [hunk 1]
```diff
@@ -1,10 +1,9 @@
 # Architecture
 
-Passing `architecture = CPU()` or `architecture = GPU()` to the `Model` constructor will determine whether the model
-is time stepped on a CPU or GPU.
+Passing `CPU()` or `GPU()` to the grid constructor determines whether the grid lives on a CPU or GPU.
 
-Ideally a set up or simulation script does not need to be modified to run on a GPU but we are still smoothing out
-rough edges. Generally the CPU wants `Array` objects while the GPU wants `CuArray` objects.
+Ideally a set up or simulation script does not need to be modified to run on a GPU but still we are smoothing
+out rough edges. Generally the CPU wants `Array` objects while the GPU wants `CuArray` objects.
```
Comment (glwagner) [inline docs/src/model_setup/architecture.md:6]: point out that users should raise issues to improve the API for CPU->GPU transitions?

File: docs/src/model_setup/callbacks.md [hunk 2]
```diff
@@ -21,30 +21,57 @@ show_time(sim) = @info "Time is $(prettytime(sim.model.clock.time))"
...
+We can also use the convenience [`add_callback!`](@ref). For example, we could add a callback like
 
-```@repl checkpointing
+```@example checkpointing
+add_callback!(simulation, show_time, name=:total_A_via_convenience, IterationInterval(2))
+
+simulation
+```
+
+By default, callbacks are called after a time-step has been completed but users
```
Comment (glwagner) [inline docs/src/model_setup/callbacks.md:41]:
```suggestion
By default, callbacks are called after a time-step has been completed. However, users
```

File: docs/src/model_setup/callbacks.md [hunk 3]
```diff
@@ -21,30 +21,57 @@ show_time(sim) = @info "Time is $(prettytime(sim.model.clock.time))"
...
+can construct callbacks that are called after tendencies are calculated, but before taking a time-step.
+The latter is useful for modifying tendency calculations.
+
+We can control when the callback is called via the `callsite` keyword argument. By default
+`callsite = TimeStepCallsite()`. If we want the callback after the tendencies are
+computed but before taking a time-step we need to construct the callback with `callsite = TendencyCallsite()`.
```
Comment (glwagner) [inline docs/src/model_setup/callbacks.md:47]:
```suggestion
By default, callbacks are called after a time-step has been completed but users
can construct callbacks that are called after tendencies are calculated, but before taking a time-step.
The latter is useful for modifying tendency calculations.

We can control when the callback is called via the `callsite` keyword argument. By default
`callsite = TimeStepCallsite()`. If we want the callback after the tendencies are
computed but before taking a time-step we need to construct the callback with `callsite = TendencyCallsite()`.
```

These paragraphs repeat information which makes them hard to understand / ambiguous. Try to condense it. Something like

```
The keyword argument `callsite` determines the moment at which the callback is executed.
By default, `callsite = TimeStepCallsite()`, indicating execution _after_ the completion of a time-step.
Two other options are... [state options, and how they can be used].
```

File: docs/src/model_setup/grids.md [hunk 4]
```diff
@@ -139,20 +139,22 @@ grid = RectilinearGrid(size = (Nx, Ny, Nz),
...
-lines!(ax1, grid.yᵃᶜᵃ[1:Ny], grid.Δyᵃᶜᵃ[1:Ny])
-scatter!(ax1, grid.yᵃᶜᵃ[1:Ny], grid.Δyᵃᶜᵃ[1:Ny])
+lines!(ax1, ynodes(grid, Center()), yspacings(grid, Center()))
+scatter!(ax1, ynodes(grid, Center()), yspacings(grid, Center()))
 
 ax2 = Axis(fig[2, 1]; xlabel = "z-spacing (m)", ylabel = "z (m)", limits = ((0, 50), nothing))
-lines!(ax2, grid.Δzᵃᵃᶜ[1:Nz], grid.zᵃᵃᶜ[1:Nz])
-scatter!(ax2, grid.Δzᵃᵃᶜ[1:Nz], grid.zᵃᵃᶜ[1:Nz])
+lines!(ax2, zspacings(grid, Center()), znodes(grid, Center()))
+scatter!(ax2, zspacings(grid, Center()), znodes(grid, Center()))
```
Comment (glwagner) [inline docs/src/model_setup/grids.md:157]: Probably using more lines for all this stuff will make it easier to read

**Review by glwagner — APPROVED** [review]

File: docs/src/model_setup/architecture.md (hunk 1 above)
Comment (navidcy) [reply docs/src/model_setup/architecture.md:6]: done

File: docs/src/model_setup/callbacks.md (hunk 3 above)
Comment (navidcy) [reply docs/src/model_setup/callbacks.md]: indeed! done

File: docs/src/model_setup/grids.md (hunk 4 above)
Comment (navidcy) [reply docs/src/model_setup/grids.md]: good point! done

**Review by jagoosw — APPROVED** [review]
Lgtm!

**Pull request merged**